### DISCUSSION TITLE
Misc

### DISCUSSION THREAD

**Discussion by ocots** [control-toolbox/LossControl.jl#17] [General]
# Remarks in vrac.

@AnasXbouali 

See the comments below.

**Comment by ocots** [comment]
- Here, add a space between to and (CS).
- In general, I would not add labels for the equations since there is no ref. There are others in other pages.

<img width="837" alt="Capture d’écran 2024-08-22 à 18 33 47" src="https://github.com/user-attachments/assets/0fbfa4b1-4257-4a46-8663-dada7a6494cc">

**Reply by AnasXbouali** [reply]
the change is made.

**Comment by ocots** [comment]
In the page [here](https://control-toolbox.org/control-loss/stable/numerical.html#Description-of-the-indirect-method), we have:

- \texttt{ForwardDiff} 
- \textit{pseudo-Hamiltonian}\footnote{the pseudo-Hamiltonian stands for the Hamiltonian flow associated with each arc.}

This should be fixed.

Still in the same place, in the paragraph:

> We begin by defining the flow of the Hamiltonian associated with each arc. To accomplish this, we use the function `Flow` that can be found in the [`CTFlows.jl`](https://github.com/control-toolbox/CTFlows.jl) package. This latter allows to solve the Hamiltonian system over a given time interval from given initial values of the state and the adjoint vector. This function requires necessary libraries such as \texttt{ForwardDiff} for calculating gradients and Jacobians and `OrdinaryDiffEq` for solving ordinary differential equations.

The usage is the following:

- for the package do not use backquotes and write the suffix: CTFlows.jl instead of `CTFlows.jl`, OrdinaryDiffEq.jl instead of `OrdinaryDiffEq`.
- for code use backquotes: `Flow` is 👍

**Reply by AnasXbouali** [reply]
the change is made.

**Comment by ocots** [comment]
[Zermelo problems](https://control-toolbox.org/control-loss/stable/zermelo.html): I would split in two different pages the two examples since it is quite long.

**Reply by AnasXbouali** [reply]
the change is made.

**Comment by ocots** [comment]
**Erratum**. It cannot be shortened as mentioned below since the dimension of the flows are different!

~~The following can be shortened:~~

```julia
ode_sol = fc((t0, tt1), x0, pp0, saveat=0.1)
ttt1 = ode_sol.t
xx1 = [ ode_sol[1:2, j] for j in 1:size(ttt1, 1) ]
pp1 = [ ode_sol[3:4, j] for j in 1:size(ttt1, 1) ]
uu1 = u11.(xx1, pp1)

pλ0 = 0.
py0 = 0.

ode_sol = fcl((tt1, tt2), [xx1[end] ; aa ; 0.0], [pp1[end] - [0. , jmp11]; pλ0 ; py0], saveat=0.1)
ttt2 = ode_sol.t
xx2 = [ ode_sol[1:2, j] for j in 1:size(ttt2, 1) ]
pp2 = [ ode_sol[5:6, j] for j in 1:size(ttt2, 1) ]
uu2 = a.*ones(length(ttt2)) ;

ode_sol = fc((tt2, tf), xx2[end], pp2[end] - [0. , jmp22], saveat=0.1)
ttt3 = ode_sol.t
xx3 = [ ode_sol[1:2, j] for j in 1:size(ttt3, 1) ]
pp3 = [ ode_sol[3:4, j] for j in 1:size(ttt3, 1) ]
uu3 = u11.(xx3, pp3)

tsol = [ ttt1 ; ttt2 ; ttt3 ]
xsol = [ xx1 ; xx2 ; xx3 ]
psol = [ pp1 ; pp2 ; pp3 ]
usol = [ uu1 ; uu2 ; uu3 ]

m = length(tsol)

x11 = [ xsol[i][1] for i=1:m ]
x22 = [ xsol[i][2] for i=1:m ]
p11 = [ psol[i][1] for i=1:m ]
p22 = [ psol[i][2] for i=1:m ]
```

You can use the concatenation of arcs. See [here](https://control-toolbox.org/OptimalControl.jl/stable/tutorial-flow.html#Concatenation-of-arcs). But you need to add the jumps. I have added this possibility. See this [example](https://github.com/control-toolbox/CTProblems.jl/blob/e6480e91f488a6bd38e03c3e7e0e3d9917563162/test/test_simple_integrator_state_and_control_constraints_nonautonomous.jl#L44) where I make a concatenation with a jump:

```julia
f = f0 * (t1, fc) * (t2, ν2, f0) # ν2 is a jump on the costate and has the same dimension as the state
```

Here `f0` is followed by `fc` whenever `t ≥ t1` which is followed by `f0` again whenever `t ≥ t2` adding a jump `ν2` to the costate. Note that you can still call `f` as any flow from any initial time to any final time.

**Reply by AnasXbouali** [reply]
Consider the Zermelo problem with two loss control regions, let us recall the Flows:
```julia
# Hamiltonian: permanent region
H1(x, u, p)  = p' * F(x, u)                 # pseudo-Hamiltonian
u11(x, p)    = atan(p[2]/p[1])              # maximizing control
Hc(x, p)     = H1(x, u11(x, p) , p )        # Hamiltonian

# Flow
fc  = Flow(Hamiltonian(Hc))

# Hamiltonian: control loss region
H2(x, λ, y, p)   = p' * F(x, λ)   + y* p' *G(λ)    # pseudo-Hamiltonian
Hcl(X, P)        = H2(X[1:2], X[3], X[4], P[1:2])  # Hamiltonian

# Flow
fcl = Flow(Hamiltonian(Hcl))
```
If I understood correctly, the concatenation of the flows should look like this:

```julia
f        = fc * (s1, jump1, fcl) * (s2, jump2, fc) * (s3, jump3, fcl) * (s4, jump4, fc)
```
However, when I run the following command:
```julia
flow_sol = f((t0, tf), x0, p0)
```
I get an error.

**Reply by ocots** [reply]
As I said, I put an erratum at the beginning of the comment. You cannot concatenate flows with different state dimensions!

**Reply by ocots** [reply]
To concatenate, you would augment the flows with smaller dimension and add constant dynamics. Not sure it will be more readable.

**Comment by ocots** [comment]
We do not use NLSolve.jl but MINPACK.jl or NonlinearSolve.jl in our documentation: see [here](https://control-toolbox.org/OptimalControl.jl/stable/tutorial-iss.html).

To see if we change here or not.

**Reply by AnasXbouali** [reply]
thanks! I used NonlinearSolve.jl.
